atomic_reactor/plugins/prepub_flatpak_create_oci.py

```diff
@@ -0,0 +1,361 @@
+        info = self.source.flatpak_json
+
+        builddir = os.path.join(self.workflow.source.workdir, "build")
+        os.mkdir(builddir)
+
+        repo = os.path.join(self.workflow.source.workdir, "repo")
+        subprocess.check_call(['ostree', 'init', '--mode=archive-z2', '--repo', repo])
+
+        runtime_id = info['runtime']
+        runtime_version = info['runtime-version']
+
+        args = {
+            'runtime_id': runtime_id,
+            'arch': get_arch(),
+            'runtime_version': runtime_version
+        }
+
+        METADATA_TEMPLATE = dedent("""\
+            [Runtime]
+            name={runtime_id}
+            runtime={runtime_id}/{arch}/{runtime_version}
+            sdk={runtime_id}/{arch}/{runtime_version}
+
+            [Environment]
+            LD_LIBRARY_PATH=/app/lib64:/app/lib
+            GI_TYPELIB_PATH=/app/lib64/girepository-1.0
+            """)
+
+        with open(os.path.join(builddir, 'metadata'), 'w') as f:
+            f.write(METADATA_TEMPLATE.format(**args))
+
+        runtime_ref = 'runtime/{runtime_id}/{arch}/{runtime_version}'.format(**args)
+
+        subprocess.check_call(['ostree', 'commit',
+                               '--repo', repo, '--owner-uid=0',
+                               '--owner-gid=0', '--no-xattrs',
+                               '--branch', runtime_ref,
+                               '-s', 'build of ' + runtime_ref,
+                               '--tree=tar=' + tarred_filesystem,
+                               '--tree=dir=' + builddir])
+        subprocess.check_call(['ostree', 'summary', '-u', '--repo', repo])
+
+        subprocess.check_call(['flatpak', 'build-bundle', repo,
+                               '--oci', '--runtime',
+                               outfile, runtime_id, runtime_version])
+
+        return runtime_ref
+
+    def _create_app_oci(self, tarred_filesystem, outfile):
+        info = self.source.flatpak_json
+        app_id = info['id']
+
+        runtime_id = info['runtime']
+        runtime_version = info['runtime-version']
+
+        builddir = os.path.join(self.workflow.source.workdir, "build")
+        os.mkdir(builddir)
+
+        repo = os.path.join(self.workflow.source.workdir, "repo")
+
+        # See comment for build_init() for why we can't use 'flatpak build-init'
+        # subprocess.check_call(['flatpak', 'build-init',
+        #                        builddir, app_id, runtime_id, runtime_id, runtime_version])
+        build_init(builddir, app_id, runtime_id, runtime_id, runtime_version)
+
+        # with gzip'ed tarball, tar is several seconds faster than tarfile.extractall
+        subprocess.check_call(['tar', 'xCfz', builddir, tarred_filesystem])
+
+        update_desktop_files(app_id, builddir)
+
+        subprocess.check_call(['flatpak', 'build-finish'] + info['finish-args'] + [builddir])
+        subprocess.check_call(['flatpak', 'build-export', repo, builddir])
+
+        # FIXME version?
```

Comment:
Is this about checking the version of the flatpak binary?

Reply:
I have no idea at this point. I might have meant that I thought that the bundle filename (outfile) should contain some version information, but it doesn't actually matter since the tarball uploaded to Koji ends up as docker-image-{id}... anyways. [Something that is on my list of things to fix after this PR lands.] Will just remove.